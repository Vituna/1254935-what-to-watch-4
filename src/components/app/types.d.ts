import {FullMoves, FilmMain} from "../../types";

interface AppStateFromStore {
  movies: FullMoves[];
  movie: FilmMain;
  authorizationStatus: string;
}

interface AppDispatchFromStore {
  login: (authData: { email: string; password: string }) => void;
}

export type AppProps = AppStateFromStore & AppDispatchFromStore;

export interface AppFromState {
  movies: FullMoves[];
}
