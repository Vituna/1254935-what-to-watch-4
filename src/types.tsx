export interface MovieListProp {
  id: number;
  title: string;
  image: string;
  preview: string;
}

export interface MoviePageProp {
  id: number;
  title: string;
  genre: string;
  runTime: string;
  year: number;
  backgroundPoster: string;
  filmPoster: string;
  rating: number;
  ratingCount: number;
  description: string;
  director: string;
  starring: string;
}

export interface Move {
  title: string;
  image: string;
  genre: string;
  year: number;
  movieDurationTime: string;
  bigPoster: string;
  rating: number;
  numberVotes: number;
  descriptionOne: string;
  descriptionTwo: string;
  director: string;
  starring: string;
  preview: string;
}

export interface FullMoves {
  id: number;
  title: string;
  filmPoster: string;
  image: string;
  backgroundPoster: string;
  backgroundColor: string;
  src: string;
  preview: string;
  description: string;
  rating: number;
  ratingCount: number;
  director: string;
  starring: string;
  runTime: string;
  genre: string;
  year: number;
}

export interface FilmMain {
  id: number;
  title: string;
  genre: string;
  year: number;
  backgroundPoster: string;
  filmPoster: string;
  isFavorites: boolean;
}

export interface Review {
  id: number;
  user: {
    id: number;
    name: string;
  };
  rating: number;
  comment: string;
  date: string;
}
