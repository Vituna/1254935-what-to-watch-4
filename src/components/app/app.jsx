import React, {PureComponent} from 'react';
import {Switch, Route, BrowserRouter} from "react-router-dom";

import Main from "../main/main.jsx";
import MoviePage from '../movie-page/movie-page.jsx';
import withTabs from '../../hocs/with-tabs.jsx';

const MoviePageWrapped = withTabs(MoviePage);

class App extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      activeCard: null,
    };

    this._handleMovieCardClick = this._handleMovieCardClick.bind(this);
  }

  _handleMovieCardClick(movieSetting) {
    this.setState({
      activeCard: movieSetting,
    });
  }

  _renderMain() {

    return (
      <Main
        onTitleClick={this._handleMovieCardClick}
        onCardClick={this._handleMovieCardClick}
      />
    );
  }

  _renderMoviePage() {

    return (
      <MoviePageWrapped
        onTitleClick={this._handleCardTitleClick}
        onCardClick={this._handleCardTitleClick}
      />
    );
  }

  _renderApp() {
    const {activeCard} = this.state;

    const isActiveCard = activeCard ? this._renderMoviePage() : this._renderMain();
    return isActiveCard;
  }

  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
          <Route exact path="/dev-film">
            {this._renderMoviePage()}
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

export default App;
