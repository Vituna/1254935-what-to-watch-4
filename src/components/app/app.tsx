import * as React from "react";
import {Switch, Route, BrowserRouter} from "react-router-dom";
import {connect} from "react-redux";

import Main from "../main/main";
import MoviePage from "../movie-page/movie-page";
import withTabs from "../../hocs/with-tabs";
import {AppProps, AppState} from "./types";

const MoviePageWrapped = withTabs(MoviePage);

class App extends React.PureComponent<AppProps, AppState> {
  constructor(props) {
    super(props);

    this.state = {
      activeCard: null,
    };

    this._handleMovieCardClick = this._handleMovieCardClick.bind(this);
  }

  _handleMovieCardClick(title) {
    this.setState({
      activeCard: title,
    });
  }

  _renderMain(): React.ReactNode {

    return (
      <Main
        onTitleClick={this._handleMovieCardClick}
        onCardClick={this._handleMovieCardClick}
      />
    );
  }

  _renderMoviePage(): React.ReactNode {
    const {movies} = this.props;
    const {activeCard} = this.state;

    return (
      <MoviePageWrapped
        movie={movies.find((film) => film.title === activeCard)}
        onTitleClick={this._handleMovieCardClick}
        onCardClick={this._handleMovieCardClick}
      />
    );
  }

  _renderApp(): React.ReactNode {
    const {activeCard} = this.state;

    const isActiveCard = activeCard ? this._renderMoviePage() : this._renderMain();
    return isActiveCard;
  }

  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
          <Route exact path="/dev-film">
            {this._renderMoviePage()}
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

const mapStateToProps: object = (state: { movies: string }) => ({
  movies: state.movies,
});

export {App};
export default connect(mapStateToProps)(App);
