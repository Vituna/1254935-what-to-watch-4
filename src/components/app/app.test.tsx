import * as React from "react";
import * as renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";

import NameSpace from "../../reducer/name-space";
import {App} from "./app";
import {FullMoves, FilmMain} from "../../types";

const mockStore = configureStore([]);

const movies: FullMoves[] = [
  {
    id: 3,
    title: `The Grand Budapest Hotel`,
    filmPoster: `img/bg-the-grand-budapest-hotel.jpg`,
    image: `img/the-grand-budapest-hotel-poster.jpg`,
    backgroundPoster: `img/the-grand-budapest-hotel-poster.jpg`,
    backgroundColor: `img/bg-the-grand-budapest-hotel.jpg`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave&apos;s friend and protege.`,
    rating: 8.9,
    ratingCount: 240,
    director: `Wes Andreson`,
    starring: `Bill Murray, Edward Norton, Jude Law, Willem Dafoe and other`,
    runTime: `1h 39m`,
    genre: `Drama`,
    year: 2014,
  }
];

const movie: FilmMain = {
  id: 2,
  title: `The Grand Budapest Hotel`,
  filmPoster: `img/bg-the-grand-budapest-hotel.jpg`,
  backgroundPoster: `img/the-grand-budapest-hotel-poster.jpg`,
  genre: `Drama`,
  year: 2014,
  isFavorites: true,
};


const createNodeMock = () => {
  return {};
};

const noop = () => {
  return;
};

it(`Render App`, () => {

  const store = mockStore({
    [NameSpace.DATA]: {
      movies,
    },
    [NameSpace.USER]: {
      authorizationStatus: `NO_AUTH`,
      favoritesFilms: [],
    },
    [NameSpace.STATE]: {
      genre: movie,
    }
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <App
            movies={movies}
            movie={movie}
            authorizationStatus={`NO_AUTH`}
            login={noop}
            isLoadingFilms={false}
            isLoadingPromoFilm={false}
          />
        </Provider>, {createNodeMock})
  .toJSON();

  expect(tree).toMatchSnapshot();
});
