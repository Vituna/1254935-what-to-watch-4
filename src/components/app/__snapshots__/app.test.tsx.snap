// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="img/the-grand-budapest-hotel-poster.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="img/bg-the-grand-budapest-hotel.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The Grand Budapest Hotel
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Drama
            </span>
            <span
              className="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/films/2/player"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="none"
                height="19"
                viewBox="0 0 19 19"
                width="19"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M0 0L19 9.5L0 19V0Z"
                  fill="#EEE5B5"
                  fillRule="evenodd"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Drama
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
